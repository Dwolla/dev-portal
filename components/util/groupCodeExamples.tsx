/* eslint-disable react/no-array-index-key */
import { Children, cloneElement } from "react";

const getLanguage = (el) =>
  el.props.originalType === "pre" &&
  (Children.toArray(el.props.children).find(
    (c: any) =>
      c.props.originalType === "code" &&
      c.props.className?.startsWith("language-")
  ) as any)?.props?.className?.replace("language-", "");

const containsCode = (el) => !!getLanguage(el);

export default function groupCodeExamples({
  into: Component,
  children,
}: {
  into: any;
  children: any;
}): JSX.Element {
  const grouped = Children.toArray(children).reduce(
    (acc: any, next: any, i) => {
      if (acc === null) {
        return [
          containsCode(next) ? (
            <Component key={i}>{next}</Component>
          ) : (
            cloneElement(next, { key: i })
          ),
        ];
      }

      const [head, last] = [acc.slice(0, acc.length - 1), acc[acc.length - 1]];

      const tail =
        last.type === Component && containsCode(next)
          ? [
              cloneElement(last, {
                children: [...Children.toArray(last.props.children), next],
              }),
            ]
          : [
              last,
              containsCode(next) ? (
                <Component key={i}>{next}</Component>
              ) : (
                cloneElement(next, { key: i })
              ),
            ];

      return [...head, ...tail];
    },
    null
  );

  return grouped as JSX.Element;
}
