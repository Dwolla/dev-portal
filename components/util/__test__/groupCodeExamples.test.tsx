import { createElement } from "react";
import groupCodeExamples from "../groupCodeExamples";

function MockMDXCreateElement(props: {
  children: any;
  originalType: string;
  id?: string;
  className?: string;
}) {
  return createElement(props.originalType, {
    children: props.children,
    id: props.id,
  });
}

function MockCodeExamples({ children: c }: any) {
  return (
    <>
      <h3>code examples</h3>
      {c}
    </>
  );
}

test("groupCodeExamples", () => {
  expect(
    groupCodeExamples({
      children: [
        <MockMDXCreateElement originalType="pre">
          <MockMDXCreateElement originalType="code" className="language-php">
            code example A
          </MockMDXCreateElement>
        </MockMDXCreateElement>,

        <MockMDXCreateElement originalType="pre">
          <MockMDXCreateElement originalType="code" className="language-java">
            code example B
          </MockMDXCreateElement>
        </MockMDXCreateElement>,

        <MockMDXCreateElement originalType="p">
          paragraph text
        </MockMDXCreateElement>,

        <MockMDXCreateElement originalType="pre">
          <MockMDXCreateElement originalType="code" className="language-php">
            code example C
          </MockMDXCreateElement>
        </MockMDXCreateElement>,

        <MockMDXCreateElement originalType="pre">
          <MockMDXCreateElement originalType="code" className="language-java">
            code example D
          </MockMDXCreateElement>
        </MockMDXCreateElement>,
      ],
      into: MockCodeExamples,
    })
  ).toMatchSnapshot();
});
