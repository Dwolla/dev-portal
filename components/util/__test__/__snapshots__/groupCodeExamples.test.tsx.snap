// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`groupCodeExamples 1`] = `
Array [
  <MockCodeExamples>
    <MockMDXCreateElement
      originalType="pre"
    >
      <MockMDXCreateElement
        className="language-php"
        originalType="code"
      >
        code example A
      </MockMDXCreateElement>
      <MockMDXCreateElement
        className="language-java"
        originalType="code"
      >
        code example B
      </MockMDXCreateElement>
    </MockMDXCreateElement>
  </MockCodeExamples>,
  <MockMDXCreateElement
    originalType="p"
  >
    paragraph text
  </MockMDXCreateElement>,
  <MockCodeExamples>
    <MockMDXCreateElement
      originalType="pre"
    >
      <MockMDXCreateElement
        className="language-php"
        originalType="code"
      >
        code example C
      </MockMDXCreateElement>
      <MockMDXCreateElement
        className="language-java"
        originalType="code"
      >
        code example D
      </MockMDXCreateElement>
    </MockMDXCreateElement>
  </MockCodeExamples>,
]
`;
